typing/typetexp.ml

```diff
@@ -96,20 +96,32 @@ let rec narrow_unbound_lid_error : 'a. _ -> _ -> _ -> _ -> 'a =
   | Longident.Lapply (flid, mlid) ->
       check_module flid;
       let fmd = Env.find_module (Env.lookup_module ~load:true flid env) env in
-      begin match Env.scrape_alias env fmd.md_type with
-      | Mty_signature _ ->
-          raise (Error (loc, env, Apply_structure_as_functor flid))
-      | Mty_alias(_, p) ->
-          raise (Error (loc, env, Cannot_scrape_alias(flid, p)))
-      | _ -> ()
-      end;
+      let mty_param_opt =
+        match Env.scrape_alias env fmd.md_type with
+        | Mty_signature _ ->
+           raise (Error (loc, env, Apply_structure_as_functor flid))
+        | Mty_alias(_, p) ->
+           raise (Error (loc, env, Cannot_scrape_alias(flid, p)))
+        | Mty_functor (_, mty_param_opt, _) -> mty_param_opt
+        | _ -> None
+      in
       check_module mlid;
-      let mmd = Env.find_module (Env.lookup_module ~load:true mlid env) env in
+      let mpath = Env.lookup_module ~load:true mlid env in
+      let mmd = Env.find_module mpath env in
       begin match Env.scrape_alias env mmd.md_type with
       | Mty_alias(_, p) ->
-          raise (Error (loc, env, Cannot_scrape_alias(mlid, p)))
-      | _ ->
-          raise (Error (loc, env, Ill_typed_functor_application lid))
+         raise (Error (loc, env, Cannot_scrape_alias(mlid, p)))
+      | mty_arg ->
+         let details =
+           match mty_param_opt with
+           | None -> None
```

Comment:
IIRC, this case only appears when the functor is generative and the argument should be `()`. So, current bug non-withstanding, we could also give a good error message here.

Reply:
It's not the only case. What can happen, and would have no details, is this:
```ocaml
module F(X : sig module type S module F : S end) = struct
  type t = X.F(Parsing).t
end
```

because we go through the `_ -> None` above, since the type of F is Mty_ident "X.S" or something.

Otherwise, yeah, we could reject functors that take (), anticipating a better behavior from the rest of the typer. In fact, fixing the typer sounds trivial.

We might be able to remove all the cases with no details.